### ISSUE TITLE
Smooth Scrolling

### ISSUE THREAD

**Issue by patrickaldis** [wezterm/wezterm#3812]
It would be great to be able to use precision scrolling when scrolling back through the terminal. i.e. scrolling 1-1 with touchpad, not just jumping back a line at a time.

This could be implemented as a configuration option - I know in a terminal emulator pixel by pixel scrolling might not be everyone's cup of tea.

**Comment by patrickaldis** [comment]
Ah my apologies, I didn't see this. Out of interest, why do you think it's out of scope for the project?

**Comment by wez** [comment]
a terminal is a fixed grid of character cells. There isn't a smooth scrolling concept for them. A terminal emulator models the behavior of that kind of terminal. Pixel level scrolling is difficult to map to that because you'd need to dynamically switch to partially showing rows from either the top or bottom depending on the direction of scrolling, and that has ramifications for the rest of the rendering machinery.

It adds complexity and I just don't feel that it is necessary.

**Comment by tmccombs** [comment]
See also https://flak.tedunangst.com/post/terminal-smooth-scrolling

**Comment by reneknuvers** [comment]
VT320 terminals had it. It looks awesome, especially on a CRT with slower phosphors. It had issues with eightfolding the redraw time, but it looked great.

Alacritty does this afak.

**Comment by theherk** [comment]
> VT320 terminals had it.

Indeed [it does look good](https://youtu.be/tSJfzrSA0ec). I think it would be a great feature.

**Comment by nh2** [comment]
Since it hasn't been said in here yet:

The main UX benefit of smooth scrolling is that it is much easier to see visually how far you're scrolling.

Check this video comparing `wezterm` with `gnome-terminal`:

https://github.com/wez/wezterm/assets/399535/be0185af-d371-43ff-8e14-725b34804a50

* In `wezterm`, it is impossible to see how far you've scrolled, leading to loss of context.
  You have to think "OK, where am I now?"
* In `gnome-terminal` it's super clear.

**Comment by Hentioe** [comment]
I recommend you try Neovide's smooth scrolling, I was impressed by it.

Install [neovide](https://github.com/neovide/neovide) + neovim, run `neovide ~/.wezterm.lua`, and feel the scrolling.

This user experience is really great.

**Comment by midirhee12** [comment]
In the UI/UX world, there's something to be said about "visual hinting" and less choppy transitions. Especially when the transitions are done repetitively (such as changing rows), the importance of smooth transitions greatly increases in importance.

There's a number of benefits and reasons for the strong use of smooth scrolling for UI/UX in both web and desktop/mobile app dev. Here's a few:
- The top and bottom rows partially show which visually hints that there's more to read without having to look at the scroll bar and lose your place in the text.
- The ability to read while scrolling. This helps when scrolling to find something in a previous command or long outputs. The user often has to stop scrolling to look when scrolling is choppy.
- Higher accessiblity (a11y) needs for reading disorders such as dyslexia. When letters float and rotate, it is necessary to visually keep track of the letters on screen to not have to re-read every scroll. Letters have to be entirely re-read every time the letters change b/c of the losing place and change of orientation of letters.


Hope this shines some light on the desire for this feature request. I think it could put wezterm in a very modern light.

**Comment by midirhee12** [comment]
> * In `gnome-terminal` it's super clear.

Also the case in `sakura` and other VTE based terminals.

**Comment by midirhee12** [comment]
I was looking into the `termwiz` source and there might be a way to move the position of the `Cell`s themselves instead of trying to move the text in-between cells which would permit pixel-based scrolling. Essentially, the position of the `Cell`s would move until they reach a full `Cell` length down. Then reset the offset and move the `Cell` down one row as usual. And then repeat for continuous pixel-based scrolling.

**Comment by midirhee12** [comment]
@wez Is it possible that you can direct me to where the `Cell` positions are set in the window. If so, I can start on a PR for this issue. I'm not very familiar with this codebase but would love to help.

**Comment by schrieveslaach** [comment]
Just curious: would that also work in conjunction with terminal multiplexers such as tmux or zellij?

**Comment by RuRo** [comment]
@schrieveslaach no, scrolling in applications like tmux, vim, zellij or any curses app is implemented by basically clearing the screen and then printing the scrolled contents again. Unless some custom dark magic is implemented to detect/interop with these applications, terminal-level smooth scrolling won't work in these apps.

On the other hand, wezterm has its own builtin multiplexing, so in a perfect world you wouldn't need tmux anyways.

**Comment by theherk** [comment]
@schrieveslaach: I thought that was what the [terminal-synchronized-output](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036) spec was meant to address, and it is implemented in zellij and tmux among other applications. This doesn't itself facilitate smooth scroll but it does allow working with single frames where some magic could be done to make it work I think. At least I have seen talk of it elsewhere.

**Comment by schrieveslaach** [comment]
Thanks for the responses. I just was wondering if there is an advanced windowing/pane/whatever protocol (like an advanced graphics protocol, see [here](https://sw.kovidgoyal.net/kitty/graphics-protocol/)) that allows applications to scroll portions of the screen smoothly (vertically and horizontally). 

You are not aware of any, right?

**Comment by theherk** [comment]
@schrieveslaach No I do not, but I think that is precisely what will be needed. Some way for applications to notify the terminal of the beginning and end for both x and y of regions where a diff for scrolling animation is required. I know very little on the topic though... just that I'm really excited for it to exist.

**Comment by alicealysia** [comment]
I may be misguided in this assumption, but surely this could be achieved with overdraw?

In effect, when a user scrolls, rather than immediately removing the existing lines and replacing them, any new lines are rendered outside the viewpane as though they were added to the windows height, these lines are not displayed on screen immediately, but rather, the entire existing viewpane shifts downwards, and the overdrawn lines are displayed as they enter the viewpane.

A good way to picture this would be as though you were working with a webpage in html, where it has an overflow option which can be set to hidden, and you move the inner element downwards. In theory this could even be achieved using some shader trickery and a framebuffer, you'd need to store the previous rect in memory, cut the new rect to only include the following lines, then stitch them together and move them downwards.

Yes, this will produce input lag when the setting is enabled, but we need to consider the actual value of smooth scrolling. It isn't for the sake of fancy animations, as fun and nice as those are, it's for the sake of accessibility and legibility. If you have dyslexia like myself, using the terminal is a genuinely painful and difficult experience. You're constantly losing your position while reading, and any accessibility fonts you use are immediately rendered useless the moment you move the viewport by a single line. It requires an entire additional section of your mind be focused entirely on the act of reading. I've literally been using neovide as a terminal emulator for the most part because while it sucks at being a terminal emulator, at least I can read it.

**Comment by parkers0405** [comment]
For this interested  I started a branch of ghostty with pixel scrolling, it also has a built in neovim-gui with pixel scrolling and animation:

https://github.com/user-attachments/assets/20820500-5da3-4333-bfec-ba410e36f955

https://github.com/parkers0405/ghostty-pixel-scroll
